sign/rpmgensig.c

```diff
@@ -276,7 +282,26 @@ static int runGPG(sigTarget sigt, const char *sigfile)
     rpmPopMacro(NULL, "__plaintext_filename");
     rpmPopMacro(NULL, "__signature_filename");
 
+    /* The child GPG process may terminate without ever opening the pipe (such
+     * as when the key is expired), causing our parent process to get forever
+     * stuck on the Fopen() call below (see fifo(7) for details).  To fix that,
+     * we need to let the underlying open(2) unblock when SIGCHLD is received,
+     * by registering a no-op handler, so here we go: */
+    struct sigaction act;
+    memset(&act, '\0', sizeof(act));
+    act.sa_handler = &gpg_sa_handler;
+    act.sa_flags = 0;  /* Make sure we don't have SA_RESTART */
+    sigaction(SIGCHLD, &act, NULL);
```

Comment:
As we're in a shared library here, we'll need to take care to save the old handler and restore it after we're done. Pay attention to error paths.

Specifically setting sa_flags to 0 seems a bit superfluous as we just memset() the struct to all zero two lines earlier. Not saying it's wrong, sometimes it's useful to signify something this way, whether that's called for here I dunno.

Reply:
Oh, restoring the handler is something that I just didn't realize at all. Thanks for that.

Yeah, it really is superfluous :) I wanted to signify that `SA_RESTART` is not set, but since we `memset()` the whole struct right above, it's really not needed and I'll remove it as well.

Reply:
OTOH there might be some other flags that we want to set, at least SA_NOCLDSTOP looks quite relevant.